File: display/activity.lua
-- The activity line, the one line the idle screen draws while a Play starts or
-- runs. It sits at the top right under the clock and names the title, so the
-- seconds a playback pod pulls and starts read as work in progress and not as a
-- screen that ignored the button.
--
-- The line holds the last title it was given after the Play ends, because it
-- fades out with the mark's energy rather than vanishing on the frame the
-- activity changes. At rest the energy is 0 and the line draws nothing.
local theme = require("theme")

local activity = {}

-- The line shares the clock's right edge and its top, so the two stack in the
-- same column. LINE_Y clears the clock's own line: the clock hangs from the
-- top margin at theme.type.small, and the gap below it keeps the two lines
-- apart without touching.
local RIGHT = theme.canvas.w - theme.margin.x
local LINE_Y = theme.margin.y + theme.type.small + 12

-- The typographic marks, as UTF-8 bytes. \226\128\156 and \226\128\157 are the
-- left and right double quotation marks, and \226\128\166 is the ellipsis. The
-- title takes real quotation marks because the line is read at a distance, and
-- the ellipsis is one character so it never wraps or spaces oddly.
local OPEN_QUOTE = "\226\128\156"
local CLOSE_QUOTE = "\226\128\157"
local ELLIPSIS = "\226\128\166"

local kind = nil
local title = nil

-- receive takes the Player's retained status as a parsed table. The title comes
-- from the Play block, which the operator resolves from the Presentation, so the
-- display formats one string and reads no series or episode fields of its own.
-- A status with no Play leaves the last title in place for the fade-out.
function activity.receive(status)
  if type(status) ~= "table" then
    return
  end
  kind = status.activity
  if type(status.play) == "table" then
    if type(status.play.title) == "string" and status.play.title ~= "" then
      title = status.play.title
    elseif type(status.play.name) == "string" and status.play.name ~= "" then
      title = status.play.name
    end
  end
end

-- activity.draw returns the line, or nil when there is nothing to say. level is
-- the energy. While a Play starts or runs the line draws opaque, whatever the
-- energy is. After that the line draws at the energy, so the ramp-down that
-- follows a reveal carries the line out with the mark's motion, and a settled
-- idle screen at energy 0 draws no line at all.
function activity.draw(level)
  if not title then
    return nil
  end
  local alpha = theme.alpha.opaque
  if kind ~= "Starting" and kind ~= "Playing" then
    if not level or level <= 0 then
      return nil
    end
    alpha = string.format("&H%02X&", math.floor(255 * (1 - level) + 0.5))
  end
  local text = "Playing " .. OPEN_QUOTE .. title .. CLOSE_QUOTE .. ELLIPSIS
  return theme.text(RIGHT, LINE_Y, text, theme.type.small, theme.color.text, 9, alpha)
end

return activity

File: idle/src/idle/activity.rs
//! The activity line: the one line the screen draws while a `Play` starts or
//! runs.
//!
//! The line draws under the clock and names the title, so the seconds a
//! playback pod pulls and starts read as work in progress and not as a screen
//! that ignored the button. The unit holds the last title after the `Play`
//! ends, because the line leaves with the mark's motion rather than vanishing
//! on the frame the activity changes.

use iced_winit::core::Point;

use super::energy;
use super::{Frame, Layout};
use crate::look;
use crate::unit::Unit;
use media_screen::status::Activity;

/// How opaque the line draws, from 0 clear to 1 full.
///
/// While a `Play` starts or runs the line is opaque, whatever the energy is.
/// The move to `Idle` starts its fade, over the same ramp down the mark
/// returns on, so the line leaves with the motion and a settled screen draws
/// no line at all.
///
/// The fade leaves from full and not from the energy the mark holds. The line
/// was full for the whole of the `Play`, and the mark may not be: a `Play`
/// that ends before it played leaves the mark part way up its ramp, and a line
/// that took that level would step down in the frame the `Play` ended.
pub fn opacity(unit: &Unit, at: f64) -> f32 {
    match unit.activity {
        Activity::Starting | Activity::Playing => 1.0,
        Activity::Idle => (1.0 - energy::ease((at - unit.ramp.since) / energy::RAMP_DOWN)) as f32,
    }
}

/// The line the screen draws for one title.
///
/// The title takes real quotation marks and one ellipsis character, because the
/// line is read from across a room.
fn line(title: &str) -> String {
    format!("Playing \u{201c}{title}\u{201d}\u{2026}")
}

/// Draw the element into the frame, in canvas units.
///
/// The line shares the clock's right edge and hangs one line pitch under it, so
/// the two read as one column and neither touches the other.
pub fn draw(frame: &mut Frame, layout: &Layout, unit: &Unit, at: f64, light: f32) {
    let Some(title) = &unit.title else {
        return;
    };
    let opacity = opacity(unit, at);
    if opacity <= 0.0 {
        return;
    }

    frame.fill_text(look::line(
        line(title),
        Point::new(layout.right(), look::MARGIN_Y + look::LINE_PITCH),
        look::Anchor::TopRight,
        look::SMALL,
        look::under(look::faded(look::text(), opacity), light),
    ));
}

#[cfg(test)]
mod tests {
    use super::*;
    use media_screen::Moment;
    use media_screen::status::{Play, Status};

    /// A unit that read one status naming a `Play`.
    fn playing(activity: Activity, at: f64) -> Unit {
        let mut unit = Unit::default();
        unit.fold(
            Moment::Status(Status {
                activity,
                play: Some(Play {
                    name: "den-tv-1".into(),
                    title: "A Film".into(),
                }),
                ..Status::default()
            }),
            at,
        );
        unit
    }

    #[test]
    fn the_line_names_the_title_in_quotation_marks() {
        assert_eq!(line("A Film"), "Playing \u{201c}A Film\u{201d}\u{2026}");
    }

    #[test]
    fn the_line_is_opaque_while_a_play_starts_or_runs() {
        assert_eq!(opacity(&playing(Activity::Starting, 0.0), 0.1), 1.0);
        assert_eq!(opacity(&playing(Activity::Playing, 0.0), 9.0), 1.0);
    }

    #[test]
    fn the_line_leaves_with_the_marks_motion() {
        let mut unit = playing(Activity::Starting, 0.0);
        unit.fold(Moment::Status(Status::default()), 1.2);

        assert_eq!(opacity(&unit, 1.2), 1.0);
        assert!(opacity(&unit, 2.45) < 1.0);
        assert_eq!(opacity(&unit, 3.7), 0.0);
    }

    /// A `Play` that ends before it played leaves the mark part way up its
    /// ramp, and the line was full through the whole of it. So the line
    /// leaves from full here too, and the frame the `Play` ended on reads
    /// the same alpha as the frame before it.
    #[test]
    fn a_play_that_ends_before_it_played_carries_the_line_out_from_full() {
        let mut unit = playing(Activity::Starting, 0.0);
        assert_eq!(opacity(&unit, 0.6), 1.0);

        unit.fold(Moment::Status(Status::default()), 0.6);

        assert_eq!(opacity(&unit, 0.6), 1.0);
        assert_eq!(opacity(&unit, 1.85), 0.5);
        assert_eq!(opacity(&unit, 3.1), 0.0);
    }

    #[test]
    fn a_settled_screen_draws_no_line() {
        let unit = playing(Activity::Idle, 0.0);
        assert_eq!(unit.title.as_deref(), Some("A Film"));
        assert_eq!(opacity(&unit, 5.0), 0.0);
    }
}
